Pyrado/pyrado/algorithms/meta/sprl.py

```diff
@@ -0,0 +1,573 @@
+        self._seed = seed
+        super().train(snapshot_mode, seed, meta_info)
+
+    def step(self, snapshot_mode: str, meta_info: dict = None):
+        self.save_snapshot()
+
+        context_mean = to.cat([spl_param.context_mean for spl_param in self._spl_parameters]).double()
+        context_cov_chol = to.cat([spl_param.context_cov_chol_flat for spl_param in self._spl_parameters]).double()
+
+        target_mean = to.cat([spl_param.target_mean for spl_param in self._spl_parameters]).double()
+        target_cov_chol = to.cat([spl_param.target_cov_chol_flat for spl_param in self._spl_parameters]).double()
+
+        for param in self._spl_parameters:
+            self.logger.add_value(f"cur context mean for {param.name}", param.context_mean.item())
+            self.logger.add_value(f"cur context cov for {param.name}", param.context_cov.item())
+
+        dim = context_mean.shape[0]
+
+        # If we are in the first iteration and have a bad performance,
+        # we want to completely reset the policy if training is unsuccessful
+        reset_policy = False
+        if self.curr_iter == 0:
+            reset_policy = True
+        until_thold_exceeded(self._performance_lower_bound * 0.3, self._max_subrtn_retries)(
+            self._train_subroutine_and_evaluate_perf
+        )(snapshot_mode, meta_info, reset_policy)
+
+        # Update distribution
+        previous_distribution = ParameterAgnosticMultivariateNormalWrapper(
+            context_mean, context_cov_chol, self._optimize_mean, self._optimize_cov
+        )
+        target_distribution = ParameterAgnosticMultivariateNormalWrapper(
+            target_mean, target_cov_chol, self._optimize_mean, self._optimize_cov
+        )
+
+        rollouts = self._subroutine.sampler.sample()
+        contexts = to.tensor(
+            np.array(
+                [
+                    [rollout.rollout_info["domain_param"][param.name] for rollout in rollouts]
+                    for param in self._spl_parameters
+                ]
+            ),
+            requires_grad=True,
+        ).T
+
+        contexts_old_log_prob = previous_distribution.distribution.log_prob(contexts.double())
+        kl_divergence = to.distributions.kl_divergence(
+            previous_distribution.distribution, target_distribution.distribution
+        )
+
+        values = to.tensor([ros.undiscounted_return() for ros in rollouts])
+
+        values = to.tensor([ros.undiscounted_return() for ros in rollouts])
+
+        values = to.tensor([ros.undiscounted_return() for ros in rollouts])
+
+        def kl_constraint_fn(x):
+            distribution = previous_distribution.from_stacked(x)
+            kl_divergence = to.distributions.kl_divergence(
+                previous_distribution.distribution, distribution.distribution
+            )
+            return kl_divergence.detach().numpy()
+
+        def kl_constraint_fn_prime(x):
+            distribution = previous_distribution.from_stacked(x)
```

Comment:
Please move this function definition outside of step(), make it private (add a leading _), and add a one line doc string """..."""

Reply:
No, this does not work. The constraints need to access things like `previous_distribution`, which are only defined in the `step()` method. Since they are not called by us we can not easily make the pure, which is why we need to keep them inside step.

Reply:
I see, please still add a doc string

Reply:
Done in efd0223227a141fe81e56cd36c0605d7f84bdc3e